drivers/crypto/crypto_mbedtls_shim.c

```diff
@@ -354,114 +284,70 @@ static int mbedtls_cipher_session_setup(const struct device *dev,
-			ret = mbedtls_aes_setkey_enc(aes_ctx,
-					ctx->key.bit_stream, ctx->keylen * 8U);
-			ctx->ops.block_crypt_hndlr = mbedtls_ecb_encrypt;
-		} else {
-			ret = mbedtls_aes_setkey_dec(aes_ctx,
-					ctx->key.bit_stream, ctx->keylen * 8U);
-			ctx->ops.block_crypt_hndlr = mbedtls_ecb_decrypt;
-		}
-		if (ret) {
-			LOG_ERR("AES_ECB: failed at setkey (%d)", ret);
-			mbedtls_aes_free(aes_ctx);
-			ret = -EINVAL;
-		}
+		session->psa_alg = PSA_ALG_ECB_NO_PADDING;
+		ctx->ops.block_crypt_hndlr = mbedtls_ecb;
 		break;
+#endif /* CONFIG_PSA_WANT_ALG_ECB_NO_PADDING */
+#if CONFIG_PSA_WANT_ALG_CBC_NO_PADDING
 	case CRYPTO_CIPHER_MODE_CBC:
-		aes_ctx = &mbedtls_sessions[ctx_idx].aes_ctx;
-		mbedtls_aes_init(aes_ctx);
-		if (op_type == CRYPTO_CIPHER_OP_ENCRYPT) {
-			ret = mbedtls_aes_setkey_enc(aes_ctx,
-					ctx->key.bit_stream, ctx->keylen * 8U);
-			ctx->ops.cbc_crypt_hndlr = mbedtls_cbc_encrypt;
-		} else {
-			ret = mbedtls_aes_setkey_dec(aes_ctx,
-					ctx->key.bit_stream, ctx->keylen * 8U);
-			ctx->ops.cbc_crypt_hndlr = mbedtls_cbc_decrypt;
-		}
-		if (ret) {
-			LOG_ERR("AES_CBC: failed at setkey (%d)", ret);
-			mbedtls_aes_init(aes_ctx);
-			ret = -EINVAL;
-		}
+		session->psa_alg = PSA_ALG_CBC_NO_PADDING;
+		ctx->ops.cbc_crypt_hndlr = mbedtls_cbc;
 		break;
+#endif /* CONFIG_PSA_WANT_ALG_CBC_NO_PADDING */
+#if CONFIG_PSA_WANT_ALG_CCM
 	case CRYPTO_CIPHER_MODE_CCM:
-		ccm_ctx = &mbedtls_sessions[ctx_idx].ccm_ctx;
-		mbedtls_ccm_init(ccm_ctx);
-		ret = mbedtls_ccm_setkey(ccm_ctx, MBEDTLS_CIPHER_ID_AES,
-					 ctx->key.bit_stream, ctx->keylen * 8U);
-		if (ret) {
-			LOG_ERR("AES_CCM: failed at setkey (%d)", ret);
-			mbedtls_ccm_free(ccm_ctx);
-			ret = -EINVAL;
-			break;
-		}
-		if (op_type == CRYPTO_CIPHER_OP_ENCRYPT) {
-			ctx->ops.ccm_crypt_hndlr = mbedtls_ccm_encrypt_auth;
-		} else {
-			ctx->ops.ccm_crypt_hndlr = mbedtls_ccm_decrypt_auth;
-		}
+		session->psa_alg = PSA_ALG_AEAD_WITH_AT_LEAST_THIS_LENGTH_TAG(PSA_ALG_CCM, 8);
+		ctx->ops.ccm_crypt_hndlr = mbedtls_aead;
 		break;
-#ifdef CONFIG_MBEDTLS_CIPHER_GCM_ENABLED
+#endif /* CONFIG_PSA_WANT_ALG_CCM */
+#if CONFIG_PSA_WANT_ALG_GCM
 	case CRYPTO_CIPHER_MODE_GCM:
-		gcm_ctx = &mbedtls_sessions[ctx_idx].gcm_ctx;
-		mbedtls_gcm_init(gcm_ctx);
-		ret = mbedtls_gcm_setkey(gcm_ctx, MBEDTLS_CIPHER_ID_AES,
-					 ctx->key.bit_stream, ctx->keylen * 8U);
-		if (ret) {
-			LOG_ERR("AES_GCM: failed at setkey (%d)", ret);
-			mbedtls_gcm_free(gcm_ctx);
-			ret = -EINVAL;
-			break;
-		}
-		if (op_type == CRYPTO_CIPHER_OP_ENCRYPT) {
-			ctx->ops.gcm_crypt_hndlr = mbedtls_gcm_encrypt_auth;
-		} else {
-			ctx->ops.gcm_crypt_hndlr = mbedtls_gcm_decrypt_auth;
-		}
+		session->psa_alg = PSA_ALG_GCM;
+		ctx->ops.gcm_crypt_hndlr = mbedtls_aead;
 		break;
-#endif /* CONFIG_MBEDTLS_CIPHER_GCM_ENABLED */
+#endif /* CONFIG_PSA_WANT_ALG_GCM */
+#endif /* CONFIG_PSA_WANT_KEY_TYPE_AES*/
 	default:
-		LOG_ERR("Unhandled mode");
-		ret = -EINVAL;
+		LOG_ERR("Unsupported mode");
+		session->in_use = false;
```

Comment:
In `mbedtls_cipher_session_free` access to `session->in_use` is performed with mutex lock - why it is not the same here?
Or other way - consider making `in_use` atomic instead and then session release will not need locking.